### PR TITLE
batched_gemm + multiple_d + gemm + multiple_d

### PR DESCRIPTION
**Pull request by ltqin**
1.This is testing version
2.Now gelu is replaced with relu for test performance
3.example:
make -j example_batched_gemm_bias_gelu_gemm_bias_xdl_fp16
4.ckProfiler:
1)example:
./build/bin/ckProfiler batched_gemm_bias_gelu_gemm_bias 0 0 1 1 0 5 1024 1024 64 128 4
2)show help
./build/bin/ckProfiler batched_gemm_bias_gelu_gemm_bias

### CODE REVIEW

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_v1.hpp [hunk 1]
```diff
@@ -0,0 +1,1231 @@
...
+                   wave_m_n_id[I1]);
+        }
+
+#endif
+        auto d0_threadwise_copy =
+            ThreadwiseTensorSliceTransfer_v2<A0B0B1DataType,
+                                             A0B0B1DataType,
+                                             decltype(d0_griddesc_m0_n0_m1_n1_m2_n2_m3_n3_n4_n5),
+                                             decltype(d0_thread_desc_m0_n0_m1_n1_m2_n2_m3_n3_n4_n5),
+                                             Sequence<I1, I1, I1, I1, I1, I1, I1, I1, I1, n4>,
```
Comment (rosenrodt) [inline include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_v1.hpp:733]: Have you tried copying bigger slice than `Sequence<I1, I1, I1, I1, I1, I1, I1, I1, I1, n4>` for each threadwise_copy.run()? That could help you avoid the quadruple nested static_for loop down below.

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_v1.hpp [hunk 2]
```diff
@@ -0,0 +1,1231 @@
...
+                                                                    b0_blockwise_copy,
+                                                                    b0_grid_buf,
+                                                                    b0_block_buf,
+                                                                    b0_block_slice_copy_step,
+                                                                    blockwise_gemm0,
+                                                                    acc0_thread_buf,
+                                                                    num_k_block_main_loop);
+            // bias+gelu
+            {
+                static_for<0, Gemm0MXdlPerWave, 1>{}([&](auto mr) {
```
Comment (rosenrodt) [inline include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_v1.hpp:890]: Ditto

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_v1.hpp [hunk 3]
```diff
@@ -0,0 +1,1231 @@
...
+
+                    block_sync_lds();
+
+                    blockwise_gemm1.Run(a1_thread_buf, b1_block_buf, c1_thread_buf);
+                }
+            } // end gemm1
+
+            a0_blockwise_copy.MoveSrcSliceWindow(a0_grid_desc_ak0_m_ak1,
+                                                 a0_block_reset_copy_step); // rewind K
+            b0_blockwise_copy.MoveSrcSliceWindow(b0_grid_desc_bk0_n_bk1,
```
Comment (rosenrodt) [inline include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_v1.hpp:987]: Where does `d0_threadwise_copy` move its slicing window along N dimension?

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_v1.hpp (hunk 1 above)
Comment (ltqin) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_v1.hpp:733]: I'll try it

Comment (ltqin) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_v1.hpp:733]: @rosenrodt From the test results, there is no obvious improvement

Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_v1.hpp:733]: Yes it's just for readability.

File: example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+static constexpr bool PadGemm0N = false;
+static constexpr bool PadGemm0K = false;
+static constexpr bool PadGemm1N = false;
+static constexpr bool PadGemm1K = false;
+
+using DeviceGemmInstance =
+    ck::tensor_operation::device::DeviceBatchedGemmBiasGeluGemmBias_Xdl_CShuffle<
+        A0Layout,
+        B0Layout,
+        ck::Tuple<D0Layout, D0Layout>,
```
Comment (asroy) [inline example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp:94]: D1Layout

File: profiler/include/profile_batched_gemm_bias_gelu_gemm_bias_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+        for(int b = 0; b < BatchCount; ++b)
+        {
+            for(int m = 0; m < M; ++m)
+            {
+                for(int n = 0; n < N; ++n)
+                {
+                    cde0_element_op(acc0_g_m_n(b, m, n), acc0_g_m_n(b, m, n), d0_g_m_n(b, m, n));
+                }
+            }
+        }
```
Comment (rosenrodt) [inline profiler/include/profile_batched_gemm_bias_gelu_gemm_bias_impl.hpp:253]:
```suggestion
        acc0_g_m_n.ForEach([&](auto& self, auto idx) { 
            cde0_element_op(self(idx), self(idx), d0_g_m_n(idx));
        }
```

File: profiler/include/profile_batched_gemm_bias_gelu_gemm_bias_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+                                                         b1_g_n_o,
+                                                         c1_g_m_o_host_result,
+                                                         a1_element_op,
+                                                         b1_element_op,
+                                                         c1_element_op);
+
+        ref_gemm1_invoker.Run(ref_gemm1_argument);
+
+        // bias
+        for(int b = 0; b < BatchCount; ++b)
```
Comment (rosenrodt) [inline profiler/include/profile_batched_gemm_bias_gelu_gemm_bias_impl.hpp:267]: Ditto

File: profiler/include/profile_batched_gemm_bias_gelu_gemm_bias_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+        auto invoker_ptr = op_ptr->MakeInvokerPointer();
+
+        if(op_ptr->IsSupportedArgument(argument_ptr.get()))
+        {
+            std::string op_name = op_ptr->GetTypeString();
+
+            float ave_time =
+                invoker_ptr->Run(argument_ptr.get(), StreamConfig{nullptr, time_kernel});
+
+            std::size_t flop      = (size_t(M) * N * K * 2 + size_t(M) * N * O * 2) * BatchCount;
```
Comment (rosenrodt) [inline profiler/include/profile_batched_gemm_bias_gelu_gemm_bias_impl.hpp:329]: Extra flops by elementwise ops may be trivial so I am fine with evaluating only GEMM+GEMM's flops. However the number of bytes from the bias arrays d0's and d1's should play a significant factor in `num_btype` (not sure if it's typo of num_byte).

File: example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp [hunk 8]
```diff
@@ -0,0 +1,508 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2022, Advanced Micro Devices, Inc. All rights reserved.
+
+/*
+Gemm + Gemm fused operation. Computes C_m_o = A_m_k * B0_k_n * B1_n_o
+                                              |------------|
+                                                   Gemm0
+                                              |---------------------|
+                                                       Gemm1
```
Comment (rosenrodt) [inline example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp:9]: Update comment string to reflect the purpose of this example

File: example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+using B0Layout  = Col;
+using D00Layout = Row;
+using D01Layout = Row;
+using B1Layout  = Row;
+using C1Layout  = Row;
+using D1Layout  = Row;
+
+// C = A * B
+// E = Relu(C + D0) + D1;
+struct AddReluAdd
```
Comment (rosenrodt) [inline example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp:60]: File name says Gelu but here we still uses Relu?

File: example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+    if(!gemm.IsSupportedArgument(argument))
+    {
+        std::cout << gemm.GetTypeString() << " does not support this problem" << std::endl;
+
+        return 0;
+    }
+
+    float ave_time = invoker.Run(argument, StreamConfig{nullptr, time_kernel});
+
+    std::size_t flop      = (size_t(M) * N * K * 2 + size_t(M) * N * O * 2) * BatchCount;
```
Comment (rosenrodt) [inline example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp:442]: Extra flops by elementwise ops may be trivial so I am fine with evaluating only GEMM+GEMM's flops. However the number of bytes from the bias arrays d0's and d1's should play a significant factor in num_btype (not sure if it's typo of num_byte).

File: profiler/src/profile_batched_gemm_bias_gelu_gemm_bias.cpp [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+using F32 = float;
+
+using Row = ck::tensor_layout::gemm::RowMajor;
+using Col = ck::tensor_layout::gemm::ColumnMajor;
+
+int profile_batched_gemm_bias_gelu_gemm_bias(int argc, char* argv[])
+{
+    enum struct GemmMatrixLayout
+    {
+        MK_NK_MN_ON_MO, // 0
```
Comment (rosenrodt) [inline profiler/src/profile_batched_gemm_bias_gelu_gemm_bias.cpp:21]: GEMM+GEMM variants has only A, B0, B1 as input and C as output. Why does there seem to have 5 matrices?

File: profiler/src/profile_batched_gemm_bias_gelu_gemm_bias.cpp [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    };
+
+    enum struct GemmDataType
+    {
+        F16_F16_F16_F16_F16, // 0
```
Comment (rosenrodt) [inline profiler/src/profile_batched_gemm_bias_gelu_gemm_bias.cpp:26]: Same as above. There should only be 4 matrices concerned.

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_v1.hpp (hunk 1 above)
Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_v1.hpp]: As discussed with @ltqin, trying bigger slice is of no perf gain and of lower priority adjustment

**Comment by rosenrodt** [comment]
Is it possible to add comparable GEMM+GEMM tests in this PR like in [test_batched_gemm_gemm_fp16.cpp](https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/test/batched_gemm_gemm/test_batched_gemm_gemm_fp16.cpp)? Considering that you've added kernel instances and profiling codes already

File: example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp (hunk 4 above)
Comment (ltqin) [reply example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp]: done

File: example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp (hunk 8 above)
Comment (ltqin) [reply example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp]: done

File: include/ck/tensor_operation/gpu/device/matrix_padder.hpp [hunk 13]
```diff
@@ -218,6 +218,165 @@ struct GemmPadder_v2
...
 };
 
+// M/N/KPerTileType could be index_t or Number<>
+template <bool PadM,
+          bool PadN,
+          bool PadK,
+          typename MPerTileType,
+          typename NPerTileType,
+          typename KPerTileType>
+struct MatrixPadder_v2
```
[the remark this answers is not in the document]
Comment (ltqin) [reply include/ck/tensor_operation/gpu/device/matrix_padder.hpp:228]: done

File: library/src/tensor_operation_instance/gpu/batched_gemm_bias_gelu_gemm_bias/device_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_f16_f16_f16_f16_gmk_gnk_gno_gmo_instance.cpp [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        DeviceBatchedGemmMultipleDGemmMultipleD_Xdl_CShuffle<     Row,      Col,   ck::Tuple<Row>,      Row,       Row,   ck::Tuple<Row>,   F16,    F16,      F32,       ck::Tuple<F16>,     F16,     F32,      F32,      F16,   ck::Tuple<F16>, PassThrough, PassThrough, CDE0ElementOp, PassThrough, PassThrough, CDE1ElementOp,    PadGemm0M,  PadGemm0N,  PadGemm0K,  PadGemm1N,  PadGemm1K,    1,        256,    256,   128,    32,   128,    32,   8,    8,    2,   32,   32,     2,     4,     4,     S<4, 64, 1>,      S<1, 0, 2>,      S<1, 0, 2>,              2,              8,              8,          true,      S<4, 64, 1>,      S<1, 0, 2>,      S<1, 0, 2>,               2,               8,               8,       true,     S< 8, 32, 1>,      S<0, 2, 1>,      S<0, 2, 1>,               1,               4,               2,      false,           1,           2,               S<1, 32, 1, 8>,               8>,
```
[the remark this answers is not in the document]
Comment (ltqin) [reply library/src/tensor_operation_instance/gpu/batched_gemm_bias_gelu_gemm_bias/device_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_f16_f16_f16_f16_gmk_gnk_gno_gmo_instance.cpp]: done

File: include/ck/tensor_operation/gpu/device/matrix_padder.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (ltqin) [reply include/ck/tensor_operation/gpu/device/matrix_padder.hpp:228]: done

File: library/src/tensor_operation_instance/gpu/batched_gemm_bias_gelu_gemm_bias/device_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_f16_f16_f16_f16_gmk_gnk_gno_gmo_instance.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (ltqin) [reply library/src/tensor_operation_instance/gpu/batched_gemm_bias_gelu_gemm_bias/device_batched_gemm_bias_gelu_gemm_bias_xdl_cshuffle_f16_f16_f16_f16_gmk_gnk_gno_gmo_instance.cpp]: done

File: profiler/include/profile_batched_gemm_bias_gelu_gemm_bias_impl.hpp (hunk 5 above)
Comment (ltqin) [reply profiler/include/profile_batched_gemm_bias_gelu_gemm_bias_impl.hpp]: done

File: example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp (hunk 8 above)
Comment (ltqin) [reply example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp]: done

File: example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp (hunk 9 above)
Comment (ltqin) [reply example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp]: done

File: profiler/src/profile_batched_gemm_bias_gelu_gemm_bias.cpp (hunk 11 above)
Comment (ltqin) [reply profiler/src/profile_batched_gemm_bias_gelu_gemm_bias.cpp]: done, +output or bias

File: profiler/src/profile_batched_gemm_bias_gelu_gemm_bias.cpp (hunk 12 above)
Comment (ltqin) [reply profiler/src/profile_batched_gemm_bias_gelu_gemm_bias.cpp]: done, +bias

File: example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp (hunk 10 above)
Comment (ltqin) [reply example/37_batched_gemm_bias_gelu_gemm_bias/batched_gemm_bias_gelu_gemm_bias_xdl_fp16.cpp]: done

File: profiler/include/profile_batched_gemm_bias_gelu_gemm_bias_impl.hpp (hunk 7 above)
Comment (ltqin) [reply profiler/include/profile_batched_gemm_bias_gelu_gemm_bias_impl.hpp]: done

**Review by asroy — APPROVED** [review]

**Pull request merged**